Add pause/resume recording functionality to Record&Play and SIP plugins

isnumanagic opened the pull request:
This PR adds pause/resume functionality to Janus recorder, and implements signaling for it in Record&Play and SIP plugins.

This functionality is useful when relaying sensitive information during recorded meeting. It is implemented by skipping writing of RTP packets during the pause duration. On resume we request PLI to fix any video artifacts that can occur due to skipped frames in recording.

lminiero commented:
I remember this was attempted already in #918, which I had several doubts on. I think your patch has a similar issue in how recordings are handled, because the way you're doing it in SIP and Record&Play this will simply introduce big holes in the recording when postprocessing: when you pause and resume after a while, packets will have huge timestamp/seq number changes, which will translate to either a ton of silence, or missing video.

isnumanagic commented:
Yes, that is by design.

isnumanagic commented:
Perhaps this can be ammended by writing silence timespans into the recording header, so that after or during post processing those segments can be skipped?

lminiero commented:
> Yes, that is by design.

I don't think that's correct. A pause/resume should be seamless in the recording too.

> Perhaps this can be ammended by writing silence timespans into the recording header, so that after or during post processing those segments can be skipped?

MJR headers can't be edited after they're written, and I don't like the idea of making a lot of changes to janus-pp-rec to accomodate this.

isnumanagic commented:
I see the packet timestamps are calculated by Janus, then the solution can be to track the total paused time and subtracting it when calculating timestamps?

lminiero commented:
The solution is to write packets to the recorder that have consistent RTP headers, as we do in other plugins e.g. for source switching and RTP forwarders.

isnumanagic commented:
Going through the code i see you use `janus_rtp_switching_context` struct for tracking timestamps and seq-numbers, and `janus_rtp_header_update` method to make changes to header. I'll try to do the same here and will update the PR.

lminiero commented:
My guess is that, once you add the context in the mix, you can just trigger its behaviour by changing its SSRC before a resume (which will trick it into thinking the source changed and so headers need to be updated).

isnumanagic commented:
Updated PR with requested changes. Just changing the ssrc did not work - I had to add time offset information.

lminiero commented:
> Updated PR with requested changes. Just changing the ssrc did not work - I had to add time offset information.

From a quick glance to the code I don't think you're using the RTP context correctly. Timestamp offset management is already done automatically by RTP contexts, when the SSRC changes: it looks to me like you're just resetting the sequence number instead, which is why you're forced to manually fix the timestamp. When recording is resumed, it's the SSRCs in the context you have to change (`a_last_ssrc`, `v_last_ssrc`), not the one in the packet: this way, even if the packet is using the same SSRC it always used, the context will think it changed, and it will trigger the rewriting logic.

isnumanagic commented:
I tried to just reset the ssrc by changing the one in the context, but as I said it does not work - the silence during the pause duration remained in the recording. This is because the base timestamp is reset, but the offset is then calcualted and applied (denoted by the `How much time since the last audio RTP packet? We compute an offset accordingly` comment). That is why i reset the seq numbering instead of entire ssrc and adjust the timestamp.

lminiero commented:
Ah I think I understand what you mean now: the SSRC change forces sequence number and timestamp to be updated, but the timestamp will take into account the time that passed, and so a lot of silence will be inserted, which shouldn't indeed happen in this case. That said, I don't think the way you're updating the timestamp offset works, and I don't like the idea of adding new properties to an already crowded structure. It should be possible to do this with the properties already there.

isnumanagic commented:
I have updated CR as requested. I have managed to remove the silence between pauses by overwriting `a/v_last_time` properties. I have removed `a/v_time_offset` properties I have added to switching context, but kept `a/v_ts_reset` flags as it is more descriptive to do `seq_reset` and `ts_reset` on resume than to change the ssrc, but I can roll back the flags as well if needed.

lminiero commented:
Thanks! I'll need time to review, since your changes modifies the code of the context update method as well, which can have a huge impact (we use it pretty much everywhere).

isnumanagic commented:
The context update method has not been functionally changed, I just extracted the `seq_reset` part (which was already there actually, i just removed duplicated code) and `ts_reset` part of ssrc reset into separate if branches for finer control.

isnumanagic commented:
So basically:
```c
if(ssrc != context->a_last_ssrc) {
  ...
  // ts_reset code
  // seq_reset code
  ...
}
```
became:
```c
if(ssrc != context->a_last_ssrc) {
  ...
  context->a_ts_reset = TRUE;
  context->a_seq_reset = TRUE;
  ...
}
if(context->a_ts_reset) {
  // ts_reset code
}
if(context->a_seq_reset) {
  // seq_reset code
}
```

isnumanagic commented:
Hey @lminiero, if you have the time please take a look.

lminiero commented:
It's IETF week so not sure when I'll be able to do a deep dive. I'll try to add some comments to the PR when I can.

lminiero commented on html/recordplaytest.js:
```diff
@@ -497,6 +497,18 @@ function startRecording() {
+				recordplay.send({message: {request: 'resume'}});
+				$('#pause').attr('hidden', false);
+				$('#resume').attr('hidden', true);
+			});
```
No need for the empty lines above. As a side note, the way you implemented this won't work for consecutive recordings: new click handlers will be added any time you start one, which means that any time you click the button it will send the same request multiple times. The way we handle this in other demos is via something like:

    .unbind('click').click(function() {

As a side note, I'd prefer the pause/resume to be a single one with a toggle (e.g., like the audio/video on/off buttons in the EchoTest demo, and others), but that's not important.

lminiero replied:
PS: we use the bootstrap `hide` class to hide/show elements. Please use that instead of the manual `hidden` attribute.

lminiero commented on plugins/janus_recordplay.c:
```diff
@@ -2007,6 +2008,32 @@ static void *janus_recordplay_handler(void *data) {
 			gateway->close_pc(session->handle);
+		} else if (!strcasecmp(request_text, "pause") || !strcasecmp(request_text, "resume")) {
+			JANUS_LOG(LOG_VERB, "Record&Play: Got pause/resume request\n");
+			int pause = !strcasecmp(request_text, "pause");
```
Why not `gboolean`?

lminiero commented on plugins/janus_recordplay.c:
```diff
@@ -2007,6 +2008,32 @@ static void *janus_recordplay_handler(void *data) {
+			JANUS_LOG(LOG_VERB, "Record&Play: Got pause/resume request\n");
+			int pause = !strcasecmp(request_text, "pause");
+			result = json_object();
+			json_object_set_new(result, "status", json_string(pause ? "paused" : "resumed"));
```
Shouldn't this check if the current session is in recording mode? It may be idle, or playing something instead: if playing, the session object will still have a `recording` object set, and you'd be sending broken events.

lminiero commented on plugins/janus_recordplay.c:
```diff
@@ -2007,6 +2008,32 @@ static void *janus_recordplay_handler(void *data) {
+					json_object_set_new(info, "event", json_string(pause ? "paused" : "resumed"));
+					if(session->recording)
+						json_object_set_new(info, "id", json_integer(session->recording->id));
+					gateway->notify_event(&janus_recordplay_plugin, session->handle, info);
```
See above: this would trigger for playback too, which is wrong.

lminiero commented on plugins/janus_sip.c:
```diff
@@ -4357,10 +4357,11 @@ static void *janus_sip_handler(void *data) {
-			if(strcasecmp(action_text, "start") && strcasecmp(action_text, "stop")) {
-				JANUS_LOG(LOG_ERR, "Invalid action (should be start|stop)\n");
+			if(strcasecmp(action_text, "start") && strcasecmp(action_text, "stop") &&
+				strcasecmp(action_text, "pause") && strcasecmp(action_text, "resume")) {
```
This needs another tab of indentation, since it's a continuation of the `if` above and is confused with the action below.

lminiero commented on plugins/janus_sip.c:
```diff
@@ -4502,6 +4503,35 @@ static void *janus_sip_handler(void *data) {
+				}
+				if(record_video || record_peer_video) {
+					gateway->send_pli(&session->handle);
+				}
```
All actions above are simple, so the brackets are unneeded.

lminiero commented on record.c:
```diff
@@ -255,6 +256,33 @@ janus_recorder *janus_recorder_create_full(const char *dir, const char *codec, c
+		return -1;
+	janus_mutex_lock_nodebug(&recorder->mutex);
+	g_atomic_int_set(&recorder->paused, TRUE);
+	janus_mutex_unlock_nodebug(&recorder->mutex);
```
There's no need for a mutex, here: it's an atomic operation. Besides, the value should be `1` and not `TRUE`.

lminiero commented on record.c:
```diff
@@ -255,6 +256,33 @@ janus_recorder *janus_recorder_create_full(const char *dir, const char *codec, c
+		recorder->context.v_seq_reset = TRUE;
+		recorder->context.v_last_time = janus_get_monotonic_time();
+	}
+	g_atomic_int_set(&recorder->paused, FALSE);
```
`FALSE` --> `0`.

lminiero commented on record.c:
```diff
@@ -384,17 +416,33 @@ int janus_recorder_save_frame(janus_recorder *recorder, char *buffer, uint lengt
 		}
 	}
+	/* Edit packet header if needed */
+	rtp_header *header = (rtp_header *)buffer;
```
`janus_rtp_header`: `rtp_header` is only there for backwards compatibility, as that's how we called it originally, which we want to get rid of.

lminiero commented on record.c:
```diff
@@ -384,17 +416,33 @@ int janus_recorder_save_frame(janus_recorder *recorder, char *buffer, uint lengt
+	timestamp = ntohl(header->timestamp);
+	if(recorder->type != JANUS_RECORDER_DATA) {
+		janus_rtp_header_update(header, &recorder->context, recorder->type == JANUS_RECORDER_VIDEO, 0);
+	}
```
Simple action, no need for brackets.

lminiero commented on record.c:
```diff
@@ -384,17 +416,33 @@ int janus_recorder_save_frame(janus_recorder *recorder, char *buffer, uint lengt
+	uint32_t ssrc = ntohl(header->ssrc);
+	uint16_t seq = ntohs(header->seq_number);
+	timestamp = ntohl(header->timestamp);
+	if(recorder->type != JANUS_RECORDER_DATA) {
```
Accessing RTP fields should be done ONLY if it's not a data recorder, so this check needs to be moved up. The way it is, if you're recording a small data message (<12 bytes) this will crash.

lminiero commented on record.c:
```diff
@@ -384,17 +416,33 @@ int janus_recorder_save_frame(janus_recorder *recorder, char *buffer, uint lengt
+			/* Restore packet header data */
+			header->ssrc = htonl(ssrc);
+			header->seq_number = htons(seq);
+			header->timestamp = htonl(timestamp);
```
See above: only for non-data.

lminiero commented on record.c:
```diff
@@ -384,17 +416,33 @@ int janus_recorder_save_frame(janus_recorder *recorder, char *buffer, uint lengt
+	/* Restore packet header data */
+	header->ssrc = htonl(ssrc);
+	header->seq_number = htons(seq);
+	header->timestamp = htonl(timestamp);
```
See above: only for non-data.

lminiero commented on record.h:
```diff
@@ -60,6 +61,10 @@ typedef struct janus_recorder {
+	/*! \brief Whether writing RTP packets is paused */
+	volatile int paused;
+	/*! \brief RTP switching context for rewriting RTP headers */
+	janus_rtp_switching_context context;
```
I'm not fond of the idea of having a context as part of the recorder itself. Our plugins all have a dedicated context which they use for recordings too, even though I guess in this case we need something that's specific to the recording itself. I'll need to think about this.

lminiero commented on record.h:
```diff
@@ -60,6 +61,10 @@ typedef struct janus_recorder {
 	volatile int header;
 	/*! \brief Whether this recorder instance can be used for writing or not */
 	volatile int writable;
+	/*! \brief Whether writing RTP packets is paused */
```
s/RTP packets/data

lminiero commented on rtp.c:
```diff
@@ -601,10 +601,17 @@ void janus_rtp_header_update(janus_rtp_header *header, janus_rtp_switching_conte
+		}
+		if(context->v_ts_reset) {
+			/* Video timestamp was paused for a while */
+			JANUS_LOG(LOG_VERB, "Video RTP timestamp reset requested");
```
This feels too verbose, maybe `LOG_HUGE` is better.

lminiero commented on rtp.c:
```diff
@@ -615,11 +622,10 @@ void janus_rtp_header_update(janus_rtp_header *header, janus_rtp_switching_conte
 		if(context->v_seq_reset) {
-			/* Video sequence number was paused for a while: just update that */
+			/* Video sequence number was paused for a while */
+			JANUS_LOG(LOG_VERB, "Video RTP sequence number reset requested");
```
See above.

lminiero commented on rtp.c:
```diff
@@ -640,10 +646,17 @@ void janus_rtp_header_update(janus_rtp_header *header, janus_rtp_switching_conte
+		}
+		if(context->a_ts_reset) {
+			/* Audio timestamp was paused for a while */
+			JANUS_LOG(LOG_VERB, "Audio RTP timestamp reset requested");
```
See above.

lminiero commented on rtp.c:
```diff
@@ -658,11 +671,10 @@ void janus_rtp_header_update(janus_rtp_header *header, janus_rtp_switching_conte
 		if(context->a_seq_reset) {
-			/* Audio sequence number was paused for a while: just update that */
+			/* Audio sequence number was paused for a while */
+			JANUS_LOG(LOG_VERB, "Audio RTP sequence number reset requested");
```
See above.

lminiero reviewed (commented):
I had another look at the changes: please notice I haven't tested this yet, which will probably not happen before next week.

isnumanagic commented:
Updated PR with requested changes. If you

isnumanagic commented:
Hey @lminiero, any updates?

lminiero commented:
@isnumanagic apologies, I went on vacation and forgot about this :hand_over_mouth: 
I'll have a look at the fixes you made at the time and let you know. Since this has an impact on how mjr files are created (the extra RTP context that is part of the code is a bug change) we'll need time to do proper testing anyway, as we use recordings a lot.

lminiero commented on plugins/janus_recordplay.c:
```diff
@@ -2007,6 +2008,33 @@ static void *janus_recordplay_handler(void *data) {
+		} else if (!strcasecmp(request_text, "pause") || !strcasecmp(request_text, "resume")) {
+			JANUS_LOG(LOG_VERB, "Record&Play: Got pause/resume request\n");
+			if(session->recording) {
+				gboolean pause = !strcasecmp(request_text, "pause");
```
I think this should probably check what the state of the recording is, before doing something. As it is, a "pause" on a paused recording will set the paused property again, and will send a new event too, which seems incorrect. I think this is particolarly important due to the way "resume" is implemented in record.c, where the last time is reset every time the method is called.

I guess the right way to do that would be a new property on the `session->recording` object: maybe a volatile int to check and tweak via `g_atomic_int_compare_and_exchange` as we do in a few other places.

lminiero commented on plugins/janus_sip.c:
```diff
@@ -4502,6 +4503,26 @@ static void *janus_sip_handler(void *data) {
+				if(record_peer_video)
+					janus_recorder_resume(session->vrc_peer);
+				if(record_video || record_peer_video)
+					gateway->send_pli(session->handle);
```
Same here, I guess, even though the property should be part of the SIP plugin session object.

lminiero commented on record.c:
```diff
@@ -255,6 +256,31 @@ janus_recorder *janus_recorder_create_full(const char *dir, const char *codec, c
+		recorder->context.v_last_time = janus_get_monotonic_time();
+	}
+	g_atomic_int_set(&recorder->paused, 0);
+	janus_mutex_unlock_nodebug(&recorder->mutex);
```
See discussion above: maybe, besides checks at the plugin level, this code should check the status of the recorder too, and not reset these properties if the recorder is not paused already.

lminiero commented on record.c:
```diff
@@ -384,17 +414,39 @@ int janus_recorder_save_frame(janus_recorder *recorder, char *buffer, uint lengt
+	/* Edit packet header if needed */
+	janus_rtp_header *header = (janus_rtp_header *)buffer;
+	uint32_t ssrc;
+	uint16_t seq;
```
These two integers should be initialized.

lminiero commented on html/recordplaytest.html:
```diff
@@ -89,7 +89,11 @@ <h3 class="panel-title">Recorder/Playout</h3>
+								<h3 class="panel-title">
+									<span id="videotitle">Remote Video</span>
+									<button class="btn-xs btn-danger pull-right" autocomplete="off" id="stop">Stop</button>
+									<button class="btn-xs btn-primary pull-right" autocomplete="off" id="pause-resume">Pause</button>
```
This button is currently visible also when playing back recordings, which shouldn't happen (especially since it does nothing in that case).

lminiero reviewed (commented):
Took me a while, but I found some time to test this: it seemed to work as expected for the brief test I made, even though I didn't push it much. I did notice some UI things, though (pause button still visible when replaying recordings), and some messaging/state things related to how pause/resume are handled that may need fixing (I'll try to make some tests for that too).

isnumanagic commented:
Uodated PR with requested changes

lminiero commented on plugins/janus_recordplay.c:
```diff
@@ -2012,25 +2014,29 @@ static void *janus_recordplay_handler(void *data) {
-					janus_recorder_resume(session->vrc);
-					janus_recorder_resume(session->drc);
-					gateway->send_pli(session->handle);
+					JANUS_LOG(LOG_VERB, "Record&Play: Tried to pause/resume recording of same state, ignoring\n");
```
I'd send the event back anyway, even if we didn't have to change the state, as there may be applications expecting an event back after sending the request.

lminiero commented on plugins/janus_sip.c:
```diff
@@ -4374,12 +4374,29 @@ static void *janus_sip_handler(void *data) {
+				g_atomic_int_get(&session->vrc_peer->paused) << 3;
+			if ((!strcasecmp(action_text, "pause") && !(~recorder_bitmap & update_bitmap)) ||
+				(!strcasecmp(action_text, "resume") && !(recorder_bitmap & update_bitmap))) {
+				JANUS_LOG(LOG_ERR, "Invalid pause/resume request (at least one of audio, video, peer_audio and peer_video should change recorder state)\n");
```
I don't think this bitmap approach works, which feels a bit convoluted. First all, this will crash Janus if you're not recording everything (e.g., just the user, just the peer, or video is not active), as in that case some recordings object would be `NULL`. Besides, I suspect this check would break if the provided requests are not aligned with the state, and go in an unrecoverable state.

Considering the state enhancement you added to `record.c` that does make the pause/resume methods idempotent, I think we can keep the code as it was in the SIP plugin for the sake of simplicity, but only send the PLI if we're really resuming (so only adding a single additional check below).

lminiero reviewed (commented):
Thanks for the quick fixes! I added a couple more notes: I think we're close to merging this :v:

isnumanagic commented:
Updated PR with requested changes

lminiero commented on plugins/janus_sip.c:
```diff
@@ -4521,26 +4504,25 @@ static void *janus_sip_handler(void *data) {
-					janus_recorder_resume(session->vrc_peer);
-				if(resume_bitmap & 10)
+				if(record_peer_video && !janus_recorder_resume(session->vrc_peer))
+					send_pli = TRUE;
```
You don't need `send_pli` here: if we're recording the peer, we need a PLI from the SIP user, not the WebRTC user. I think you can just do a

	if(record_video && !janus_recorder_resume(session->vrc))
		gateway->send_pli(session->handle);

without the additional `send_pli` boolean definition.

lminiero commented:
LGTM now, thanks for all the fixes! Pinging @atoppi to give a review as well (in case I missed anything), and @alexamirante to check if this might cause issues with the way we handle recordings in production.

atoppi commented:
this looks good, I just don't get why we need the `ts_reset` flag, is this just a design choice?

atoppi reviewed (commented):
left my only question as a separate comment

isnumanagic commented:
@atoppi It was a design choice. It's more intuitive to set ts_reset and seq_reset flags when resuming recording, than to manually change SSRC, expecting it to reset timestamp and sequence number.

atoppi commented:
ok, thanks or clarifying, LGTM :+1:

lminiero commented:
Ack, merging then :+1:  Thanks for the patience!

lminiero commented:
@isnumanagic you created a monster: I'm way too often using the pause/resume feature just to do funny faces! I see a potential for doing TikTok like videos with WebRTC :rofl:
